File: frontend/src/api/sigma.ts
import { api } from "./client";

export interface SigmaCompileResponse {
  ok: boolean;
  query: string | null;
  title: string | null;
  description: string | null;
  error: string | null;
}

export interface SigmaTestSampleHit {
  timestamp: string | null;
  host_id: string | null;
  event_id: string | null;
  process: Record<string, unknown> | null;
  file: Record<string, unknown> | null;
}

export interface SigmaTestResponse {
  query: string;
  total: number;
  samples: SigmaTestSampleHit[];
}

export const sigmaApi = {
  compile: (body: string) =>
    api<SigmaCompileResponse>("/api/sigma/compile", { method: "POST", body: { body } }),
  testAdhoc: (body: string, lookbackHours = 24) =>
    api<SigmaTestResponse>("/api/sigma/test", {
      method: "POST",
      body: { body, lookback_hours: lookbackHours },
    }),
  testSavedRule: (ruleId: string, body: string | null = null, lookbackHours = 24) =>
    api<SigmaTestResponse>(`/api/sigma/rules/${ruleId}/test`, {
      method: "POST",
      body: { body, lookback_hours: lookbackHours },
    }),
};

File: backend/app/schemas/sigma.py
"""Sigma compile + test payloads."""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class SigmaCompileRequest(BaseModel):
    body: str = Field(min_length=1)


class SigmaCompileResponse(BaseModel):
    ok: bool
    query: str | None = None
    title: str | None = None
    description: str | None = None
    error: str | None = None


class SigmaTestRequest(BaseModel):
    """Body for POST /api/rules/{id}/test or /api/rules/test (ad-hoc)."""

    body: str | None = Field(
        default=None, description="Override the saved rule body for this run only."
    )
    lookback_hours: int = Field(default=24, ge=1, le=168)


class SigmaTestSampleHit(BaseModel):
    timestamp: str | None
    host_id: str | None
    event_id: str | None
    process: dict[str, Any] | None
    file: dict[str, Any] | None


class SigmaTestResponse(BaseModel):
    query: str
    total: int
    samples: list[SigmaTestSampleHit]
